new(test): add `test/vm` for localhost VM-based driver kernel compatibility tests

incertum opened the pull request:
Signed-off-by: Melissa Kilby <melissa.kilby.oss@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

/kind documentation

> /kind failing-test

> /kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-udig

> /area libscap

> /area libpman

> /area libsinsp

/area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

Edited:

Add `test/vm` (localhost automated VM based kernel - compiler version grid-search driver tests) to `test` targets similar to `e2e` sinsp tests.

CI integration is tracked here https://github.com/falcosecurity/libs/issues/531.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
new(test): add `test/vm` for localhost VM-based driver kernel compatibility tests 
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/524#" title="Author self-approved">incertum</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign andreagit97 for approval by writing `/assign @andreagit97` in a comment. For more information see:[The Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["andreagit97"]} -->

FedeDP commented:
Just WOW 😲
This is something we have been needing for a while! 
Having something similar in the CI is super valuable indeed!
Perhaps a cmake target to run the tests would help a bit :)
I will give it a look asap, at least for some initial feedback!
Thank you very much for this contribution!

jasondellaluce commented:
This is real good Melissa! We've been looking for something like this for a while now. Thank you a lot!

incertum commented:
> Just WOW 😲 This is something we have been needing for a while! Having something similar in the CI is super valuable indeed! 

`localhost` obviously is more tractable to gain more confidence in your code changes and spot issues you care about in your own deployment :) which typically is only a few distros and a smaller grid of kernel releases ...

Proper CI needs to care about more which is such a daunting task, that's probably why we are where we are.
https://github.com/falcosecurity/libs/issues/531

> Perhaps a cmake target to run the tests would help a bit :) I will give it a look asap, at least for some initial feedback! Thank you very much for this contribution!

Let me look into this more as this gets a bit more polished. Will also look more into https://github.com/falcosecurity/libs/pull/506 to converge to a good common style for such tests.

Andreagit97 commented:
Sorry, I'm just trying the CI :) 
/retest

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/524#" title="Author self-approved">incertum</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign andreagit97 for approval by writing `/assign @andreagit97` in a comment. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["andreagit97"]} -->

incertum commented:
@FedeDP: Added cmake targets as I finally got to refactoring and polishing it a bit. Thanks again for the suggestion, it helped making the driver-sanity tests cleaner and easier to use.
@Molter73: Tried to make the cmake setup similar to `e2e` targets. Thanks to your work on e2e it was pretty straight forward as I used the e2e cmake setup as template.
@LucaGuerra: Thanks for your feedback in the CI integration discussion. Did not forget about making sure to reboot the VM after each kmod test to be resilient against buggy kernel modules.

Removed WIP as v1 would now be considered complete. The PR is now ready for review. Aware that it's a random collection of scripts, mostly bash scripts. Happy to address any style feedback throughout in an attempt to make the scripts easier to follow and maintain, I think there is improvement potential in this regard.

Outlook: Emulation support still missing and arm64 support as well ... let's first see if v1 is useful for x86_64 for now.

Few observations: Noticed that in general `ubuntu` distro appears to be the most difficult one. Re-ran these tests many times and a few ubuntu kmod tests sometimes randomly just did not work. `centos7` stable in comparison.
Unrelated to this PR also noticed that the eBPF verifier on ubuntu let's minor bugs pass compared to other distros (same kernel versions), hence yum based distros appear to be the better choice for dev for sure. Also kmod never seemed to work for newer ubuntu kernels (e.g. 5.19 or 6.0), while centos7 was fine.

FedeDP commented:
/milestone 0.11.0

poiana commented:
Issues go stale after 90d of inactivity.

Mark the issue as fresh with `/remove-lifecycle stale`.

Stale issues rot after an additional 30d of inactivity and eventually close.

If this issue is safe to close now please do so with `/close`.

Provide feedback via https://github.com/falcosecurity/community.

/lifecycle stale

incertum commented:
/remove-lifecycle stale

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/524#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented:
Suggestions for first pass review:

- Test if it even works on `localhost` for you, please read README carefully
- Scripts are glue-code type of scripting, mostly bash, nothing very polished (motto: 80-20 wins the day)
- Posting my test run results here https://github.com/falcosecurity/libs/issues/982
- ubuntu VM loop is more flaky now than when I last worked on it, but centos7 VM loop is more stable -> please also read the README for additional comments and suggestions
- Using old fashioned ssh remote commands and scp as vagrant rsync and other tricks would sometimes not work.
- modern_bpf would be added in a follow up PR

Molter73 commented on test/driver_sanity/kernel_compat/scripts/vagrant_loop.sh:
```diff
@@ -0,0 +1,92 @@
+  ${SSH_BASE_COMMAND} "sudo bash /home/vagrant/vagrant_change_kernel.sh";
+
+  printf "\n\n\n\nSleeping for 45 seconds ...\n\n\n\n";
+  sleep 45;
```
Same as above w.r.t. sleep

Molter73 commented on test/driver_sanity/kernel_compat/scripts/vagrant_scap_open_test.sh:
```diff
@@ -0,0 +1,39 @@
+
+echo "Test run for compiler/kernel ${COMPILER}/$(uname -r)"
+
+rm -f /tmp/o;
```
Looks like we might not care about preserving this file, if we are rebooting the machine between calls to this script, why not use a temporary file instead?
```suggestion
OUTPUT_FILE="$(mktemp)"
```

Molter73 commented on test/driver_sanity/kernel_compat/scripts/vagrant_scap_open_test.sh:
```diff
@@ -0,0 +1,39 @@
+rm -f /tmp/o;
+# background scap-open
+if [[ ${COMPILER} == *"clang"* ]]; then
+  /home/vagrant/scap-open --bpf "/home/vagrant/driver/${COMPILER}/$(uname -r).o" > /tmp/o &
```
I think we could change this to print the output as it is written to the file instead of doing a `cat` afterwards, WDYT?
```suggestion
  /home/vagrant/scap-open --bpf "/home/vagrant/driver/${COMPILER}/$(uname -r).o" | tee /tmp/o &
```

Molter73 commented on test/driver_sanity/kernel_compat/scripts/vagrant_scap_open_test.sh:
```diff
@@ -0,0 +1,39 @@
+elif [[ ${COMPILER} == *"gcc"* ]]; then
+  cp -f "/home/vagrant/driver/${COMPILER}/$(uname -r).ko" /home/vagrant/scap.ko;
+  insmod /home/vagrant/scap.ko; lsmod | grep scap;
+  /home/vagrant/scap-open --kmod > /tmp/o &
```
Same as previous comment

Molter73 commented on test/driver_sanity/kernel_compat/scripts/vagrant_scap_open_test.sh:
```diff
@@ -0,0 +1,39 @@
+# SIGINT to force STDOUT be captured in file for verification
+pkill -SIGINT scap-open;
+cat /tmp/o;
+sleep 1;
```
Could we instead wait till the process is killed?

Molter73 commented on test/driver_sanity/kernel_compat/scripts/vagrant_scap_open_test.sh:
```diff
@@ -0,0 +1,39 @@
+cat /tmp/o;
+sleep 1;
+
+TEST=$(cat /tmp/o | grep 'captured' | sed "s/[^0-9]//g");
```
[ultra-nit]
```suggestion
TEST=$(grep 'captured' /tmp/o | sed "s/[^0-9]//g");
```

Molter73 commented on test/driver_sanity/kernel_compat/vm_provider/virtualbox/Vagrantfile:
```diff
@@ -0,0 +1,20 @@
+    m.vm.box = "bento/ubuntu-20.04"
+    m.vm.hostname = "ubuntu"
+    m.vbguest.auto_update = false
+    m.vm.network "private_network", ip: "192.168.56.26"
```
Same as previous comment
```suggestion
    m.vm.network "private_network", ip: ENV["LIBS_UBUNTU2004_IP"] || "192.168.56.26"
```

Molter73 commented on test/driver_sanity/kernel_compat/kernels.txt:
This looks like a json lines file, could we rename it to `kernels.jsonl`?

Molter73 reviewed (commented):
This is awesome @incertum, I still need to run the tests but they look like they take a pretty long time, I'll try to do so ASAP.

I apologize for the large amount of comments but, in fairness, this is a pretty big change itself.

leogr commented on test/driver_sanity/kernel_compat/scripts/kernel_plot_compat_matrix.py:
```diff
@@ -0,0 +1,94 @@
+
+def arg_parser():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--driver_artifacts_dir', help='Directory that contains subfolders by compiler version that contain the driver build artifact')
+    parser.add_argument('--save_png_dir', help='Directory to save png output file into')
```
Strongly agree :+1:

leogr commented on test/driver_sanity/kernel_compat/scripts/main.go:
```diff
@@ -0,0 +1,103 @@
+	if err != nil {
+		panic(err)
+	}
+	fmt.Printf("Performing concurrent docker run for compiler %s and compiler version %d\n", compilerType, verNumeric)
```
I was wondering... why we can't just use golang for everything (instead of python or bash) :angel:

incertum replied:
let me re-audit this part and also respond w/ a general comment re a possible re-design @leogr with more Go.

incertum commented on test/driver_sanity/kernel_compat/scripts/main.go:
yes bash background or anything didn't work and I think we save a lot of time this way, it finishes in 2 min vs maybe 10-12 min?

incertum commented on test/driver_sanity/kernel_compat/scripts/vagrant_change_kernel.sh:
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
+if [[ "${EUID}" -ne 0 ]]; then
+  echo "Usage: sudo bash vagrant_change_kernel.sh"
```
could be vagrant inside VM thing, let me check.

incertum commented on test/driver_sanity/kernel_compat/scripts/vagrant_loop.sh:
```diff
@@ -0,0 +1,92 @@
+        cp "../../build/driver/${compiler_version}/${kernel_uname_r}.o" "../../build/driver-ok/${compiler_version}/${kernel_uname_r}.o";
+      elif [[ ${compiler_version} == *"gcc"* ]]; then
+        cp "../../build/driver/${compiler_version}/${kernel_uname_r}.ko" "../../build/driver-ok/${compiler_version}/${kernel_uname_r}.ko";
+        ${SSH_BASE_COMMAND} "sudo reboot"; sleep 10;
```
`vagrant reload` is a great idea, let me check if this helps w/ making ubuntu reboots into a new kernel more resilient as this is currently very flaky (mostly ubuntu not centos7)

incertum commented:
Thanks @Molter73 for the very detailed first sweep of comments :heart:, will address all of them. Re @leogr https://github.com/falcosecurity/libs/pull/524#discussion_r1141376936 ... convert to Go would first like to discuss this more before pushing next changes ...


Pros:
- Some sort of Go module would fit better into existing testing framework of the broader project
- Would get rid of some bash helper scripts and/or the loops in a subset of bash scripts
- Likely cleaner code than pure bash scripts w/ limitations listed below in mind


Cons:
- Turning this into a Go module could be regarded as glorified bash script as we rely a lot on shell core-utils, so would shelling out a lot from Go be ok? We would have to ...


WDYT @leogr and @Molter73 I would be fine either way, keep bash or try to replacing as much as possible w/ Go. Just not a big fan of Python :upside_down_face: with some exceptions, please read more below:


Python:

We have just one little Python script for the "Data Science" part of this. Python has the strongest packages for these type of quick data transformations on small data plus all the fancy plotting libraries. Not sure if we can get rid of this and use Go instead ... not aware of many Data Science wrangling and graphing use cases or good libraries in this regard in Go.

Molter73 commented:
> WDYT @leogr and @Molter73 I would be fine either way, keep bash or try to replacing as much as possible w/ Go.

I'm fine with either honestly, but if most of the Go code is just gonna be creating commands in strings to then be run by a separate shell, I would just keep the bash scripts, just with a little more linting and formatting to prevent them from becoming a hellish nightmare (as bash tends to do).

As for the little python in these changes, I know people like to talk against python nowadays (yes its performance is usually garbage, yes it has some concurrency issues, yes it is not a strongly typed language), but I honestly like it for some things, it's great for data science and our e2e tests without the fixture feature from pytest would be a nightmare (and I honestly haven't seen this approach on any other language/test framework). So, to me, it's fine to keep the one python script in the PR.

incertum commented:
Rebased and staged some things we needed for testing, haven't gotten to addressing all change requests yet @Molter73 .

incertum commented on test/vm/kernels.jsonl:
```diff
@@ -0,0 +1,12 @@
+{"headers": ["http://vault.centos.org/centos/6.9/os/x86_64/Packages/kernel-devel-2.6.32-696.el6.x86_64.rpm"], "kernels": ["http://vault.centos.org/centos/6.9/os/x86_64/Packages/kernel-2.6.32-696.el6.x86_64.rpm"]}
+{"headers": ["http://mirror.centos.org/centos/7.9.2009/updates/x86_64/Packages/kernel-devel-3.10.0-1160.49.1.el7.x86_64.rpm"], "kernels": ["http://mirror.centos.org/centos/7.9.2009/updates/x86_64/Packages/kernel-3.10.0-1160.49.1.el7.x86_64.rpm"]}
+{"headers": ["http://amazonlinux.us-east-1.amazonaws.com/2/core/2.0/x86_64/3cea6107e9fffdc60f8ca99c7dac421b427d2ea06003cf5eb970344e4c2e18f4/../../../../../blobstore/13b711e5117ca3ad3c459a303dedfdb6dc0ad7c1440aabf02df6407a43bd94df/kernel-devel-4.14.299-223.520.amzn2.x86_64.rpm"], "kernels": [""]}
```
@FedeDP changing URL to get the kernels RPMs work for everything in the kernel crawler output except for AmazonLinux2 and the other newer AmazonLinux ones and we would need those to make the test robust. Would you have ideas re how to retrieve the correct kernel RPM, the URL is quite strange ...

FedeDP replied:
> Would you have ideas re how to retrieve the correct kernel RPM, the URL is quite strange ...

Well, I expected `kernel`, or `linux`, or `linux-image` to be present; but they are not...

A `yum search` run on amazonlinux 2023 ec2 instance gives:
```
sudo yum search kernel
Last metadata expiration check: 0:52:55 ago on Mon Jun  5 12:32:10 2023.
========================================================================================= Name & Summary Matched: kernel =========================================================================================
kernel.x86_64 : The Linux kernel
kernel-devel.x86_64 : Development package for building kernel modules to match the kernel
kernel-headers.x86_64 : Header files for the Linux kernel for use by glibc
```

but headers and kernel are not there...

FedeDP replied:
Well, it seems like some `kernel`s are there; i updated `kernel-crawler` amazonlinux2 impl to return all deps, and it returned:
```
...
{
"kernelversion": 1,
"kernelrelease": "4.19.84-33.70.amzn2.x86_64",
"target": "amazonlinux2",
"headers": [
"http://amazonlinux.us-east-1.amazonaws.com/2/extras/kernel-ng/stable/x86_64/f9bb600b110c667c6e2c57454f861bb81462dc01d366747d6614565287e7f066/../../../../../../blobstore/87c3cff795dd362a9788483354a0e82165136ef6cf851df525f749414ca7f3de/kernel-devel-4.19.84-33.70.amzn2.x86_64.rpm"
]
},
{
"kernelversion": 1,
"kernelrelease": "4.19.84-33.70.amzn2.x86_64",
"target": "amazonlinux2",
"headers": [
"http://amazonlinux.us-east-1.amazonaws.com/2/extras/kernel-ng/stable/x86_64/f9bb600b110c667c6e2c57454f861bb81462dc01d366747d6614565287e7f066/../../../../../../blobstore/dcfb3c9186b315da366f3d82ac77bdc9d77b86020cc1c10a5262546225249511/kernel-4.19.84-33.70.amzn2.x86_64.rpm"
]
},
{
"kernelversion": 1,
"kernelrelease": "4.19.72-25.58.amzn2.x86_64",
"target": "amazonlinux2",
"headers": [
"http://amazonlinux.us-east-1.amazonaws.com/2/extras/kernel-ng/stable/x86_64/f9bb600b110c667c6e2c57454f861bb81462dc01d366747d6614565287e7f066/../../../../../../blobstore/652cee0fc7c4dd8036aa5e6488a20277bdb7a4ffcaa896b9748ddb9110dd5979/kernel-devel-4.19.72-25.58.amzn2.x86_64.rpm"
]
},
{
"kernelversion": 1,
"kernelrelease": "4.19.72-25.58.amzn2.x86_64",
"target": "amazonlinux2",
"headers": [
"http://amazonlinux.us-east-1.amazonaws.com/2/extras/kernel-ng/stable/x86_64/f9bb600b110c667c6e2c57454f861bb81462dc01d366747d6614565287e7f066/../../../../../../blobstore/907a972944ed2eb1b7e58793c399f398520729ac6b17506d037acb5c2eeec371/kernel-4.19.72-25.58.amzn2.x86_64.rpm"
]
}
]
```
and so on.

FedeDP replied:
Also,
```
grep "4.14.299" out.txt
"kernelrelease": "4.14.299-223.520.amzn2.x86_64",
"http://amazonlinux.us-east-1.amazonaws.com/2/core/2.0/x86_64/e5756b881dd10f379f337f77c404dbc4c557dcbacab9b1b8a3aea0fb859ea0fb/../../../../../blobstore/01ff4f0b1a4f8a68cd1fdf8723581bf3eeb5e70cb31cc49294e33ec841d5e09f/kernel-4.14.299-223.520.amzn2.x86_64.rpm"
"kernelrelease": "4.14.299-223.520.amzn2.x86_64",
"http://amazonlinux.us-east-1.amazonaws.com/2/core/2.0/x86_64/e5756b881dd10f379f337f77c404dbc4c557dcbacab9b1b8a3aea0fb859ea0fb/../../../../../blobstore/13b711e5117ca3ad3c459a303dedfdb6dc0ad7c1440aabf02df6407a43bd94df/kernel-devel-4.14.299-223.520.amzn2.x86_64.rpm"
```

FedeDP replied:
So, they are stored in different blob stores that's why just updating the name is not working.
I'll share the small patch with you, basically `to_driverkit_config` should become
```python
def to_driverkit_config(self, release, deps):
    return [repo.DriverKitConfig(release, "amazonlinux{2,2022,2023}", dep) for dep in deps]
```

for each amazonlinux supported version.
I'll attach the patch [here](https://github.com/falcosecurity/libs/files/11653099/patch.txt).

incertum replied:
Works like a charm, ty ❤️ !

incertum reviewed (commented):
@Molter73 know you currently have limited availability, just letting you know I addressed your feedback. Thanks for a couple of new bash tricks to make the VM tests more robust, most notably using `vagrant reload` tremendously helped to recover from the kmod tests.


@FedeDP  tested on my fedora and ubuntu Linux boxes and macOS (Intel chip). Fingers crossed it also works on other distros. `modern_bpf` has to wait until I check on our build containers, last time I checked about 3 months ago I would have needed to add yet another custom container to build scap-open w/ modern_bpf ... and we don't want that and rather use the official falco-builder.

incertum commented:
Rebased and shared current test results here https://github.com/falcosecurity/libs/issues/982#issuecomment-1581918966

maxgio92 commented on test/vm/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+)
+
+add_custom_target(vm-container
+	COMMAND time docker build -f ${CMAKE_CURRENT_SOURCE_DIR}/containers/ubuntu2004.Dockerfile -t ${UBUNTU_CONTAINER1} .;
```
Just a note: when `time` is not installed as binary, the shell `sh` `time` builtin seems not to be picked, ending up to:
```
time: no such file or directory
```
Experienced and fixed On Arch Linux, by installing separately time package.

maxgio92 commented on test/vm/containers/debianbuster.Dockerfile:
```diff
@@ -0,0 +1,145 @@
+LABEL maintainer="cncf-falco-dev@lists.cncf.io"
+# adopted from https://raw.githubusercontent.com/falcosecurity/driverkit/master/docker/builders/builder-any-x86_64_gcc8.0.0_gcc6.0.0_gcc5.0.0_gcc4.9.0_gcc4.8.0.Dockerfile
+
+ARG TARGETARCH
```
I suppose we'd need a default value, otherwise all the deb packages won't have valid URL, like `https://download.falco.org/dependencies/cpp-6_6.3.0-18_${TARGETARCH}.deb`.
Experienced a build failure on Arch Linux and Docker client 24.0.2, where the resulting *cpp-6* package's URL is `https://download.falco.org/dependencies/cpp-6_6.3.0-18_.deb`, leading to an unexpected response body when running later `dpkg -i`. WDYT?

maxgio92 replied:
```suggestion
ARG TARGETARCH=amd64
```

maxgio92 commented on test/vm/kernels.jsonl:
```diff
@@ -0,0 +1,11 @@
+{"headers": ["http://vault.centos.org/centos/6.9/os/x86_64/Packages/kernel-devel-2.6.32-696.el6.x86_64.rpm"], "kernels": ["http://vault.centos.org/centos/6.9/os/x86_64/Packages/kernel-2.6.32-696.el6.x86_64.rpm"]}
+{"headers": ["http://mirror.centos.org/centos/7.9.2009/updates/x86_64/Packages/kernel-devel-3.10.0-1160.49.1.el7.x86_64.rpm"], "kernels": ["http://mirror.centos.org/centos/7.9.2009/updates/x86_64/Packages/kernel-3.10.0-1160.49.1.el7.x86_64.rpm"]}
+{"headers": ["http://amazonlinux.us-east-1.amazonaws.com/2/core/2.0/x86_64/e5756b881dd10f379f337f77c404dbc4c557dcbacab9b1b8a3aea0fb859ea0fb/../../../../../blobstore/635c3af49a948c3844fc8b638ef25898aeeb863ea647a417e6e8a4f2d58fc231/kernel-devel-4.14.296-222.539.amzn2.x86_64.rpm"], "kernels": ["http://amazonlinux.us-east-1.amazonaws.com/2/core/2.0/x86_64/e5756b881dd10f379f337f77c404dbc4c557dcbacab9b ...
+{"headers": ["https://mirror.rcg.sfu.ca/mirror/ELRepo/kernel/el7/x86_64/RPMS/kernel-lt-devel-5.4.231-1.el7.elrepo.x86_64.rpm"], "kernels": ["https://mirror.rcg.sfu.ca/mirror/ELRepo/kernel/el7/x86_64/RPMS/kernel-lt-5.4.231-1.el7.elrepo.x86_64.rpm"]}
```
I've found missing Centos kernel headers packages, from mirrors:

- https://mirror.rcg.sfu.ca/mirror/ELRepo/kernel/el7/x86_64/RPMS/kernel-lt-devel-5.4.231-1.el7.elrepo.x86_64.rpm -> 503
- https://mirror.rcg.sfu.ca/mirror/ELRepo/kernel/el7/x86_64/RPMS/kernel-ml-devel-5.10.16-1.el7.elrepo.x86_64.rpm
- -> 503
- https://mirror.rcg.sfu.ca/mirror/ELRepo/kernel/el7/x86_64/RPMS/kernel-ml-devel-5.14.15-1.el7.elrepo.x86_64.rpm -> 503
- http://mirrors.coreix.net/elrepo/kernel/el7/x86_64/RPMS/kernel-ml-devel-6.3.4-1.el7.elrepo.x86_64.rpm -> 404

detailed log:
```
--2023-06-17 13:38:13--  https://mirror.rcg.sfu.ca/mirror/ELRepo/kernel/el7/x86_64/RPMS/kernel-lt-devel-5.4.231-1.el7.elrepo.x86_64.rpm
Resolving mirror.rcg.sfu.ca (mirror.rcg.sfu.ca)... 142.58.94.6
Connecting to mirror.rcg.sfu.ca (mirror.rcg.sfu.ca)|142.58.94.6|:443... connected.
HTTP request sent, awaiting response... 503 Service Unavailable
2023-06-17 13:38:14 ERROR 503: Service Unavailable.

--2023-06-17 13:38:14--  https://mirror.rcg.sfu.ca/mirror/ELRepo/kernel/el7/x86_64/RPMS/kernel-ml-devel-5.10.16-1.el7.elrepo.x86_64.rpm
Reusing existing connection to mirror.rcg.sfu.ca:443.
HTTP request sent, awaiting response... 503 Service Unavailable
2023-06-17 13:38:14 ERROR 503: Service Unavailable.

--2023-06-17 13:38:14--  https://mirror.rcg.sfu.ca/mirror/ELRepo/kernel/el7/x86_64/RPMS/kernel-ml-devel-5.14.15-1.el7.elrepo.x86_64.rpm
Reusing existing connection to mirror.rcg.sfu.ca:443.
HTTP request sent, awaiting response... 503 Service Unavailable
2023-06-17 13:38:14 ERROR 503: Service Unavailable.

--2023-06-17 13:38:14--  http://mirrors.coreix.net/elrepo/kernel/el7/x86_64/RPMS/kernel-ml-devel-6.3.4-1.el7.elrepo.x86_64.rpm
Resolving mirrors.coreix.net (mirrors.coreix.net)... 85.13.241.50, 2a01:c0:2:3d::2
Connecting to mirrors.coreix.net (mirrors.coreix.net)|85.13.241.50|:80... connected.
HTTP request sent, awaiting response... 404 Not Found
2023-06-17 13:38:14 ERROR 404: Not Found.
```

Maybe we'd like to find alternative ones?

incertum replied:
Amazing, yes why would we ever something just a little bit stable ... adjusted it, but more importantly left some notes in the readme that mirrors could be unavailable or break over time ...

Andreagit97 commented:
tried it on Ubuntu 22.04 `Linux  5.19.0-45-generic #46~22.04.1-Ubuntu x86_64` and it worked without issues :tada:! It took me almost 1 hour but I've not a powerful setup (8 CPUs) or a powerful bandwidth (~10MBit/s on lucky days)

incertum commented:
> tried it on Ubuntu 22.04 `Linux 5.19.0-45-generic #46~22.04.1-Ubuntu x86_64` and it worked without issues 🎉! It took me almost 1 hour but I've not a powerful setup (8 CPUs) or a powerful bandwidth (~10MBit/s on lucky days)

Thanks Andrea for reporting back, yes this sounds about right when you do everything from scratch (building containers, extracting kernel headers etc). Re-launching tests should go faster, and at the end it's a perfect test you do on the side to check before pushing major kernel driver changes up or when you are too lazy for manual tests 🙃 

Max is still checking if it works on ArchLinux. Meanwhile anyone having more change requests wrt to the bash scripts or other aspects? Planning to build it out (add modern_bpf) and improve over time, hopefully we are agreeing on what a good first version is soon. Thanks!

maxgio92 commented on test/vm/vm_provider/virtualbox/Vagrantfile:
```diff
@@ -0,0 +1,33 @@
+Vagrant.configure("2") do |config|
+
+  config.vm.provider "virtualbox" do |v|
```
Maybe it's good to ensure the `vagrant-vbguest` is installed @incertum? WDYT?

incertum replied:
Done and also added a general dependency check for all conditions I could think of atm ...

maxgio92 replied:
Thank you @incertum!!

maxgio92 commented on test/vm/README.md:
````diff
@@ -0,0 +1,349 @@
+
+```bash
+# Check dependencies
+make vm-dependency-check;
````
Hey @incertum, after running the commands above I don't find this target. Should it require some steps furter?

maxgio92 commented:
Hi all,
I just finished the testing on Arch Linux and I confirm it worked:
![image](https://github.com/falcosecurity/libs/assets/7593929/7fe0af8d-b63d-4064-8330-b486b0bc24f0)

incertum commented:
Thanks a ton for assisting in testing this @maxgio92 🙏 !

incertum commented:
@falcosecurity/libs-maintainers what needs to happen to get this PR merged?

This PR addresses the localhost testbed for developers as outlined in https://github.com/falcosecurity/libs/blob/master/proposals/20230530-driver-kernel-testing-framework.md#test-infrastructure and has proven useful on multiple occasions.

Furthermore, I would propose to defer major improvements or refactors to future PRs to also open up the possibility for others to help improve it. What do you think?

poiana commented:
LGTM label has been added.  <details>Git tree hash: e769b66a9136d603c536cdd4838609ae7769456e</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/524#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/524#pullrequestreview-1506951099" title="Approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum,jasondellaluce]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Molter73 commented on test/vm/scripts/compile_scap_open.sh:
```diff
@@ -0,0 +1,29 @@
+  exit 1
+fi
+
+tar -xvf ${LIBS_TAR_GZ} -C ${LIBS_DIR}/;  # fresh extraction of libs src in container, clean build dir
```
Same as a previous comment, don't you need to `rm -rf "${LIBS_DIR}"` to ensure the directory is clean?

Molter73 commented on test/vm/scripts/kernel_download.sh:
```diff
@@ -0,0 +1,33 @@
+  rm -rf ${HEADERS_DIR}; mkdir -p "${HEADERS_DIR}";
+  rm -rf ${KERNELS_DIR}; mkdir -p "${KERNELS_DIR}";
+
+  cat "${KERNELS_FILE}" | jq -r '.headers[]' | sed 's/\"//g' > /tmp/headers
```
Is the sed part removing quotes? Is this needed? I ran the commands locally and the `-r` parameter on jq should take care of it.

Molter73 commented on test/vm/scripts/kernel_download.sh:
```diff
@@ -0,0 +1,33 @@
+  cat "${KERNELS_FILE}" | jq -r '.headers[]' | sed 's/\"//g' > /tmp/headers
+  wget -i /tmp/headers --directory-prefix=${HEADERS_DIR} 
+
+  cat "${KERNELS_FILE}" | jq -r '.kernels[]' | sed 's/\"//g' > /tmp/kernels
```
Same as previous comment.

Molter73 reviewed (commented):
Sorry for the delay on this @incertum!! I left a few more comments, but overall LGTM and unless someone else has objections I think we can go ahead and merge.

poiana commented:
New changes are detected. LGTM label has been removed.

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/524#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/524#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented:
Thank you @Molter73 for taking another look, appreciate it :pray: !

Also re-worked the previous commit of adding more info to the readme here https://github.com/falcosecurity/libs/pull/524/commits/3d96343644b635a97125f13d1a378074fee5cead more clearly stating how this is different and separate from the approach we will take for the new kernel driver CI-powered tests. I thought it was important to highlight as there will be no overlap in setup, scripts and design choices, because of these very different use cases. CC @LucaGuerra @therealbobo @FedeDP

incertum commented:
@Molter73 kindly asking if you have additional change requests for v1 or if we can proceed? Thanks!

incertum commented:
Hi everyone, kind bump, thank you!

Molter73 reviewed (approved):
/lgtm

Sorry for the delay on this, LGTM!

poiana commented:
LGTM label has been added.  <details>Git tree hash: 45f77f49c9b008e201597b72bfc6bc55a2bcc778</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/524#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/524#pullrequestreview-1524566982" title="LGTM">Molter73</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Molter73,incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/524#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/524#pullrequestreview-1527756392" title="Approved">jasondellaluce</a>*, *<a href="https://github.com/falcosecurity/libs/pull/524#pullrequestreview-1524566982" title="LGTM">Molter73</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Molter73,incertum,jasondellaluce]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented:
/milestone 5.1.0+driver
